src/dict.cpp

```diff
@@ -230,6 +231,144 @@ int dictRehash(dict *d, int n) {
     return 1;
 }
 
+dictAsyncRehashCtl *dictRehashAsyncStart(dict *d, int buckets) {
+    if (!dictIsRehashing(d)) return 0;
+
+    d->asyncdata = new dictAsyncRehashCtl(d, d->asyncdata);
+
+    int empty_visits = buckets * 10;
+
+    while (d->asyncdata->queue.size() < (size_t)buckets && d->rehashidx < d->ht[0].size) {
+        dictEntry *de;
+
+        /* Note that rehashidx can't overflow as we are sure there are more
+         * elements because ht[0].used != 0 */
+        while(d->ht[0].table[d->rehashidx] == NULL) {
+            d->rehashidx++;
+            if (--empty_visits == 0) goto LDone;
+            if (d->rehashidx >= d->ht[0].size) goto LDone;
+        }
+
+        de = d->ht[0].table[d->rehashidx];
+        // We have to queue every node in the bucket, even if we go over our target size
+        while (de != nullptr) {
+            d->asyncdata->queue.emplace_back(de);
+            de = de->next;
+        }
+        d->rehashidx++;
+    }
+
+LDone:
+    if (d->asyncdata->queue.empty()) {
+        // We didn't find any work to do (can happen if there is a large gap in the hash table)
+        auto asyncT = d->asyncdata;
+        d->asyncdata = d->asyncdata->next;
+        delete asyncT;
+        return nullptr;
+    }
+    return d->asyncdata;
+}
+
+void dictRehashAsync(dictAsyncRehashCtl *ctl) {
+    for (size_t idx = ctl->hashIdx; idx < ctl->queue.size(); ++idx) {
+        auto &wi = ctl->queue[idx];
+        wi.hash = dictHashKey(ctl->dict, dictGetKey(wi.de));
+    }
+    ctl->hashIdx = ctl->queue.size();
+    ctl->done = true;
+}
+
+bool dictRehashSomeAsync(dictAsyncRehashCtl *ctl, size_t hashes) {
+    size_t max = std::min(ctl->hashIdx + hashes, ctl->queue.size());
+    for (; ctl->hashIdx < max; ++ctl->hashIdx) {
+        auto &wi = ctl->queue[ctl->hashIdx];
+        wi.hash = dictHashKey(ctl->dict, dictGetKey(wi.de));
+    }
+
+    if (ctl->hashIdx == ctl->queue.size()) ctl->done = true;
+    return ctl->hashIdx < ctl->queue.size();
+}
+
+void dictCompleteRehashAsync(dictAsyncRehashCtl *ctl, bool fFree) {
+    dict *d = ctl->dict;
```

Comment:
Should there be a check for ctl != nullptr at the beginning of these functions?

Reply:
Calling this with a NULL arg is a programming error, and avoiding it will only cause harder to debug issues upstream.  NULL crash is pretty quick and immediate to track down.